File: Mesh/fbinary_search.m
function [i_inf]=fbinary_search(x,x0)
    % x a sorted vector (typically a grid vector)
    % Performs binary search and return the largest index such that x(i) <= x0
    if(nargin==0); unit_test(); return; end

    i_inf=1;
    i_sup=numel(x);

    % Safety test
    if(x0<x(1))
        i_inf=-1;
        return
    end
    if(x0>=x(end))
        i_inf=i_sup;
        return
    end


    % We loop until we narrow down to one index
    while (i_inf+1<i_sup)
        mid=(floor((i_inf+i_sup)/2));
        if (x(mid)<=x0)
            i_inf=mid;
        else
            i_sup=mid;
        end
    end
end

function unit_test()
    fbinary_search(1:10,1)   == 1
    fbinary_search(1:10,10)  == 10
    fbinary_search(1:10,11)  == 10
    fbinary_search(1:10,0)   == -1
    fbinary_search(1:10,5.1) == 5
    fbinary_search(1:10,5)   == 5
end

File: mechlib/mesh/fbinary_search.py
import numpy as np
    
def fbinary_search(x = None,x0 = None): 
    # x a sorted vector (typically a grid vector)
# Performs binary search and return the largest index such that x(i) <= x0
    if (len(varargin) == 0):
        unit_test()
        return i_inf
    
    i_inf = 1
    i_sup = np.asarray(x).size
    # Safety test
    if (x0 < x(1)):
        i_inf = - 1
        return i_inf
    
    if (x0 >= x(end())):
        i_inf = i_sup
        return i_inf
    
    # We loop until we narrow down to one index
    while (i_inf + 1 < i_sup):

        mid = (int(np.floor((i_inf + i_sup) / 2)))
        if (x(mid) <= x0):
            i_inf = mid
        else:
            i_sup = mid

    
    return i_inf
    
    
def unit_test(): 
    fbinary_search(np.arange(1,10+1),1) == 1
    fbinary_search(np.arange(1,10+1),10) == 10
    fbinary_search(np.arange(1,10+1),11) == 10
    fbinary_search(np.arange(1,10+1),0) == - 1
    fbinary_search(np.arange(1,10+1),5.1) == 5
    fbinary_search(np.arange(1,10+1),5) == 5
    return
    
    return i_inf